pypose/optim/optimizer.py

```diff
@@ -1,39 +1,244 @@
+            self.corrector = Trivial() if corrector is None else corrector
+
+    @torch.no_grad()
+    def step(self, inputs, targets=None):
+        r'''
+        Performs a single optimization step.
+
+        Args:
+            inputs (Tensor/LieTensor or tuple of Tensors/LieTensors): the inputs to the model.
+            targets (Tensor/LieTensor): the model targets to approximate.
+                If not given, the model outputs are minimized. Defaults: ``None``.
+
+        Return:
+            Tensor: the minimized model loss, i.e.,
+            :math:`\sum_i \rho( \|\bm{f}(\bm{\theta},\bm{x}_i)-\bm{y}_i)\|^2)`.
+
+        Note:
+            Different from PyTorch optimizers like
+            `SGD <https://pytorch.org/docs/stable/generated/torch.optim.SGD.html>`_, where the model
+            error has to be a scalar, the output of model :math:`\bm{f}` can be a Tensor/LieTensor or a
+            tuple of Tensors/LieTensors.
+
+            See more details of
+            `Gauss-Newton (GN) algorithm <https://en.wikipedia.org/wiki/Gauss-Newton_algorithm>`_ on
+            Wikipedia.
+
+        Example:
+            Optimizing a simple module to **approximate pose inversion**.
+
+            >>> class PoseInv(nn.Module):
+            ...     def __init__(self, *dim):
+            ...         super().__init__()
+            ...         self.pose = pp.Parameter(pp.randn_se3(*dim))
+            ...
+            ...     def forward(self, inputs):
+            ...         # the last dimension of the output is 6,
+            ...         # which will be the residual dimension.
+            ...         return (self.pose.Exp() @ inputs).Log()
+            ...
+            >>> posinv = PoseInv(2, 2)
+            >>> inputs = pp.randn_SE3(2, 2)
+            >>> optimizer = pp.optim.GN(posinv)
+            ...
+            >>> for idx in range(10):
+            ...     error = optimizer.step(inputs)
+            ...     print('Pose Inversion error %.7f @ %d it'%(error, idx))
+            ...     if error < 1e-5:
+            ...         print('Early Stopping with error:', error.item())
+            ...         break
+            ...
+            Pose Inversion error: 1.6865690 @ 0 it
+            Pose Inversion error: 0.1065131 @ 1 it
+            Pose Inversion error: 0.0002673 @ 2 it
+            Pose Inversion error: 0.0000005 @ 3 it
+            Early Stopping with error: 5.21540641784668e-07
+        '''
+        R = self.model(inputs, targets)
+        for pg in self.param_groups:
+            numels = [p.numel() for p in pg['params'] if p.requires_grad]
+            J = modjac(self.model, inputs=(inputs, targets), flatten=True)
+            R, J = self.corrector(R = R, J = J)
+            D = self.solver(A = J, b = -R.view(-1, 1)).split(numels)
+            [p.add_(d.view(p.shape)) for p, d in zip(pg['params'], D) if p.requires_grad]
+        return self.model.loss(inputs, targets)
```

Comment:
self.model is required to have the loss() function. So it is not a simple nn.module.
The same comment for the LM.

Reply:
the loss() function is from the standardized model, not the model that user provide any more: see this line

https://github.com/pypose/pypose/blob/cca97237eb0120e31d8acf50097b70a8a400a0ef/pypose/optim/optimizer.py#L141

Reply:
I could change the `loss` method to `_loss` if you think this name may conflict with the users' self-defined method.